core/src/main/java/org/infinispan/interceptors/impl/PrefetchInterceptor.java

```diff
@@ -207,68 +214,97 @@ private Object prefetchKeysIfNeededAndInvokeNext(InvocationContext ctx, Visitabl
                nil -> makeStage(prefetchKeysIfNeededAndInvokeNext(ctx, command, keys, true)).toCompletableFuture()));
       }
       if (transferedKeys != null) {
-         return asyncInvokeNext(ctx, command, retrieveRemoteValues(ctx, transferedKeys).toCompletableFuture());
+         return asyncInvokeNext(ctx, command, retrieveRemoteValues(ctx, transferedKeys, command.getTopologyId()).toCompletableFuture());
       } else {
          return invokeNext(ctx, command);
       }
    }
 
-   private InvocationStage retrieveSingleRemoteValue(InvocationContext ctx, Object key) {
+   private InvocationStage retrieveSingleRemoteValue(InvocationContext ctx, Object key, int topologyId) {
+      if (trace) {
+         log.tracef("Locally prefetching entry for key %s", key);
+      }
       GetCacheEntryCommand getCacheEntryCommand = commandsFactory.buildGetCacheEntryCommand(key, EnumUtil.bitSetOf(Flag.CACHE_MODE_LOCAL));
-      return makeStage(invokeNextThenApply(ctx, getCacheEntryCommand, (ctx1, command1, rv) -> {
-         CacheEntry entry = (CacheEntry) rv;
-         Metadata metadata = entry != null ? entry.getMetadata() : null;
-         CompletableFuture<InternalCacheValue> future;
-         if (metadata != null && metadata.version() != null && svm.isVersionActual(keyPartitioner.getSegment(key), metadata.version())){
-            return null;
-         } else if ((metadata instanceof RemoteMetadata)) {
-            Address backup = ((RemoteMetadata) metadata).getAddress();
-            future = retrieveRemoteValue(Collections.singleton(backup), key);
-         } else {
-            future = retrieveRemoteValue(null, key);
+      // We want to throw OTE with the proper requested topology early if the command's topology does not match.
+      // Moreover, OTE thrown this way has requested topology id set; all queries about owner etc. must match
+      // command's topology id. If it did not the cache topology could change between SDI and STI
+      // (when the exception pops up) and we would set next topology id too high, while the OTE/AOLE was thrown
+      // due to older topology being present in SDI.
```

Comment:
I'd rather point the reader to the `ScatteredDistributionInterceptor` (?) method that checks the topology id and throws the OTE instead. This looks like too much detail and too far from the implementation to stay true for long.